### PR TITLE
Add validation to `Address`

### LINKED ISSUE #176: Add sanity check validation of `Address`es
The `Address` class is simply a wrapper of a `String` and doesn't perform any validation on the form of the `String` to make sure it is an actual XRPL address. As seen in #175, this lack of validation can cause errors to happen while signing/doing other things with the `Address` which could be caught earlier on (and more gracefully) by validating the structure of the address string.

I propose the following validation in `Address`:
1) Make sure the address starts with `r`
2) Check that the length of the String is 34 characters

### PR DESCRIPTION
**Pull request by mukulljangid**
fix #176

### CODE REVIEW

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Wrappers.java [hunk 1]
```diff
@@ -36,6 +36,16 @@ public String toString() {
...
     }
 
+    /**
+     * Validates that a {@link Address}'s value's length is equal to 34 characters and starts with `r`.
+     */
+    @Value.Check
+    public void validateAddress() {
+      Preconditions.checkArgument(this.value().startsWith("r"),"Invalid Address: Bad Prefix");
+      Preconditions.checkArgument(this.value().length() >= 25 && this.value().length() <= 35,
+        "Classic Addresses must be 34 characters long.");
```
Comment (sappenin) [inline xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Wrappers.java:46]: This error message seems out of sync with the Precondition.

Comment (sappenin) [inline xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Wrappers.java:46]: Also, let's add a unit test to validate the error message (unless I missed it somewhere below?)

**Review by sappenin — CHANGES_REQUESTED** [review]

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Wrappers.java (hunk 1 above)
Comment (mukulljangid) [reply xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Wrappers.java]: I'll add one

Comment (mukulljangid) [reply xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/transactions/Wrappers.java]: I'll change it!

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java [hunk 2]
```diff
@@ -17,7 +17,7 @@
 import java.util.Optional;
 
 @SuppressWarnings( {"ParameterName", "MethodName", "LocalVariableName"})
-public class XAddressTest {
+public class AddressTest {
```
Comment (nkramer44) [inline xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java:20]: Why the rename here? This should stay `XAddressTest` and should only test the `XAddress` class. Any tests that test `Address` should go in another test class called `AddressTest`

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java [hunk 3]
```diff
@@ -253,4 +253,51 @@ public void xAddressWith64BitTag() {
...
+  }
+
+  @Test
+  public void addressOfIncorrectLength() {
+
+    assertThrows(
+      IllegalArgumentException.class,
+      () -> {
+        Address address = Address.of("r9cZA1mLK5R");
+        addressCodec.classicAddressToXAddress(address, true);
```
Comment (nkramer44) [inline xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java:289]: This line will never be reached.

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java [hunk 4]
```diff
@@ -253,4 +253,51 @@ public void xAddressWith64BitTag() {
...
+      },
+      "Classic Addresses must be (25,35) characters long inclusive."
+    );
+    assertThrows(
+      IllegalArgumentException.class,
+      () -> {
+        Address address = Address.of("rAJYB9cZA1mLK5R5Am25ArfXFmqgNwjZgnfk59");
+        addressCodec.classicAddressToXAddress(address, true);
```
Comment (nkramer44) [inline xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java:297]: This line will never be reached.

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java [hunk 5]
```diff
@@ -253,4 +253,51 @@ public void xAddressWith64BitTag() {
...
+  @Test
+  public void addressWithBadChecksum() {
+    Address address = Address.of("r9cZA1mLK5R5am25ArfXFmqgNwjZgnfk59");
+
+    assertThrows(
+      EncodingFormatException.class,
+      () -> addressCodec.classicAddressToXAddress(address, true),
+      "Checksum does not validate"
+    );
+  }
```
Comment (nkramer44) [inline xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java:266]: This test should go in `AddressCodecTest`. The above XAddress tests should also probably go in there. Can you create a ticket to move those over?

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java [hunk 6]
```diff
@@ -253,4 +253,51 @@ public void xAddressWith64BitTag() {
...
+
+
+  @Test
+  public void addressWithBadPrefix() {
+
+    assertThrows(
+      IllegalArgumentException.class,
+      () -> {
+        Address address = Address.of("c9cZA1mLK5R5Am25ArfXFmqgNwjZgnfk59");
+        addressCodec.classicAddressToXAddress(address, true);
```
Comment (nkramer44) [inline xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java:276]: This line will never be reached.

**Review by nkramer44 — CHANGES_REQUESTED** [review]

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+      "Classic Addresses must be (25,35) characters long inclusive."
+    );
+  }
+
+  @Test
+  public void useValidateAddress() {
+    Address address1 = Address.of("r9cZA1mLK5R5Am25ArfXFmqgNwjZgnfk59");
+    assertDoesNotThrow(() -> {
+      address1.validateAddress();
+    });
```
Comment (nkramer44) [inline xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java:50]: No need to test this method directly, it will get called on the `Address.of()` method.

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+
+  @Test
+  public void addressWithBadPrefix() {
+
+    assertThrows(
+      IllegalArgumentException.class,
+      () -> {
+        Address address = Address.of("c9cZA1mLK5R5Am25ArfXFmqgNwjZgnfk59");
+      },
```
Comment (nkramer44) [inline xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java:21]:
```suggestion
      () -> Address.of("c9cZA1mLK5R5Am25ArfXFmqgNwjZgnfk59"),
```

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  @Test
+  public void addressOfIncorrectLength() {
+
+    assertThrows(
+      IllegalArgumentException.class,
+      () -> {
+        Address address = Address.of("r9cZA1mLK5R");
+      },
```
Comment (nkramer44) [inline xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java:33]:
```suggestion
      () -> Address.of("r9cZA1mLK5R"),
```

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+      "Classic Addresses must be (25,35) characters long inclusive."
+    );
+    assertThrows(
+      IllegalArgumentException.class,
+      () -> {
+        Address address = Address.of("rAJYB9cZA1mLK5R5Am25ArfXFmqgNwjZgnfk59");
+      },
```
Comment (nkramer44) [inline xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java:40]:
```suggestion
      () -> Address.of("rAJYB9cZA1mLK5R5Am25ArfXFmqgNwjZgnfk59"),
```

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/XAddressTest.java [hunk 11]
```diff
@@ -220,35 +219,23 @@ public void convertBetweenClassicAndXAddressTestnet(
...
-      () -> addressCodec.xAddressToClassicAddress(xAddress),
+      () -> addressCodec.xAddressToClassicAddress(xaddress),
       "Invalid X-Address: Bad Prefix"
     );
   }
 
   @Test
-  public void xAddressWith64BitTag() {
-    XAddress xAddress = XAddress.of("XVLhHMPHU98es4dbozjVtdWzVrDjtV18pX8zeUygYrCgrPh");
+  public void xaddressWith64BitTag() {
```
Comment (nkramer44) [inline xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/XAddressTest.java:233]: Why the rename here?

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressCodecTest.java [hunk 12]
```diff
@@ -153,6 +155,29 @@ public void encodeDecodeAccountPublicKey() {
...
+  @Test
+  public void xaddressWithBadChecksum() {
+    XAddress xaddress = XAddress.of("XVLhHMPHU98es4dbozjVtdWzVrDjtV5fdx1mHp98tDMoQXa");
+
+    assertThrows(
+      EncodingFormatException.class,
+      () -> addressCodec.xAddressToClassicAddress(xaddress),
+      "Checksum does not validate"
+    );
+  }
```
Comment (nkramer44) [inline xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressCodecTest.java:178]: Sorry, I got confused. I didn't realize `XAddressTest` was in the address codec module. Can you move these back to `XAddressTest` like they were before?

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
+package org.xrpl.xrpl4j.codec.addresses;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import org.junit.jupiter.api.Test;
+import org.xrpl.xrpl4j.model.transactions.Address;
+
+public class AddressTest {
```
Comment (nkramer44) [inline xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java:9]: I think the reason this isn't getting picked up by codecov is because it's in the wrong module. These tests should go in the `xrpl4j-model` module because these don't test any classes in the address codec module.

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+  private final AddressCodec addressCodec = new AddressCodec();
```
Comment (nkramer44) [inline xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java:11]: Can remove this.

**Review by nkramer44 — CHANGES_REQUESTED** [review]

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java (hunk 7 above)
Comment (mukulljangid) [reply xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java:50]: I just added this to check if the codecov changes with this. I'll remove this.

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/XAddressTest.java (hunk 11 above)
Comment (mukulljangid) [reply xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/XAddressTest.java:233]: It was not following the regex in `AddressCodecTest` so I changed here as well

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java (hunk 14 above)
Comment (mukulljangid) [reply xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/AddressTest.java:11]: Got added back, since I moved the checksum test back here!

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/XAddressTest.java (hunk 11 above)
Comment (nkramer44) [reply xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/XAddressTest.java:233]: Can you revert this, and add this to `AddressCodecTest`?

```
@SuppressWarnings( {"ParameterName", "MethodName", "LocalVariableName"})
public class AddressCodecTest {
```

Comment (mukulljangid) [reply xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/XAddressTest.java:233]: Okay!

**Review by nkramer44 — APPROVED** [review]

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/XAddressTest.java [hunk 15]
```diff
@@ -246,7 +233,7 @@ public void xAddressWithBadPrefix() {
   public void xAddressWith64BitTag() {
     XAddress xAddress = XAddress.of("XVLhHMPHU98es4dbozjVtdWzVrDjtV18pX8zeUygYrCgrPh");
 
-    assertThrows(
+    org.junit.jupiter.api.Assertions.assertThrows(
```
Comment (sappenin) [inline xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/XAddressTest.java:236]: Plz revert this change.

```suggestion
    assertThrows(
```

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/XAddressTest.java [hunk 16]
```diff
@@ -215,27 +214,15 @@ public void convertBetweenClassicAndXAddressTestnet(
...
-    );
-  }
-
-
   @Test
   public void xAddressWithBadPrefix() {
     XAddress xAddress = XAddress.of("dGzKGt8CVpWoa8aWL1k18tAdy9Won3PxynvbbpkAqp3V47g");
 
-    assertThrows(
+    org.junit.jupiter.api.Assertions.assertThrows(
```
Comment (sappenin) [inline xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/XAddressTest.java:225]: Plz revert this change.

```suggestion
    assertThrows(
```

File: xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/XAddressTest.java [hunk 17]
```diff
@@ -178,7 +177,7 @@ public void convertBetweenClassicAndXAddressMainnet(
       .test(false)
       .build();
 
-    assertThat(fromXAddress).isEqualTo(classicAddress);
+    Assertions.assertThat(fromXAddress).isEqualTo(classicAddress);
```
Comment (sappenin) [inline xrpl4j-address-codec/src/test/java/org/xrpl/xrpl4j/codec/addresses/XAddressTest.java:180]: Can we unwind all the changes that add `Assertions.` - they're just adding noise, and are out of step with our style through most of the other tests.

**Review by sappenin — CHANGES_REQUESTED** [review]
Great work -- we're _so close_!!

Just a few minor changes and this will be good to merge.

**Review by sappenin — APPROVED** [review]
